crates/nu-cli/tests/completions/mod.rs

```diff
@@ -292,6 +294,112 @@ fn customcompletions_fallback() {
+        def "foo test bar" [] {}"#;
+    assert!(support::merge_input(command.as_bytes(), &mut engine, &mut stack).is_ok());
+
+    let mut completer = NuCompleter::new(Arc::new(engine), Arc::new(stack));
+    let completion_str = "foo test";
+    let suggestions = completer.complete(completion_str, completion_str.len());
+    // including both subcommand and directory completions
+    let expected: Vec<String> = vec!["foo test bar".into(), folder("test_a"), folder("test_b")];
+    match_suggestions(&expected, &suggestions);
+}
+
+/// Custom function flags mixed with subcommands
+#[test]
+fn custom_flags_and_subcommands() {
+    let (_, _, mut engine, mut stack) = new_engine();
+    let command = r#"
+        def foo [--test: directory] {}
+        def "foo --test bar" [] {}"#;
+    assert!(support::merge_input(command.as_bytes(), &mut engine, &mut stack).is_ok());
+
+    let mut completer = NuCompleter::new(Arc::new(engine), Arc::new(stack));
+    let completion_str = "foo --test";
+    let suggestions = completer.complete(completion_str, completion_str.len());
+    // including both flag and directory completions
+    let expected: Vec<String> = vec!["foo --test bar".into(), "--test".into()];
+    match_suggestions(&expected, &suggestions);
+}
+
+/// Only complete subcommands if argument type is something like int/string
+#[test]
+fn custom_arguments_vs_subcommands() {
+    let (_, _, mut engine, mut stack) = new_engine();
+    let command = r#"
+        def foo [i: string] {}
+        def "foo test bar" [] {}"#;
+    assert!(support::merge_input(command.as_bytes(), &mut engine, &mut stack).is_ok());
+
+    let mut completer = NuCompleter::new(Arc::new(engine), Arc::new(stack));
+    let completion_str = "foo test";
+    let suggestions = completer.complete(completion_str, completion_str.len());
+    // including both subcommand and directory completions
+    let expected: Vec<String> = vec!["foo test bar".into()];
+    match_suggestions(&expected, &suggestions);
+}
+
+/// External command only if starts with `^`
+#[test]
+#[cfg(not(windows))]
+fn external_commands_only() {
+    let (_, _, mut engine, stack) = new_engine();
+    engine.add_env_var(
+        "PATH".into(),
+        Value::List {
+            vals: vec![Value::String {
+                val: "/bin/".into(),
+                internal_span: Span::unknown(),
+            }],
+            internal_span: Span::unknown(),
+        },
+    );
+    let mut completer = NuCompleter::new(Arc::new(engine), Arc::new(stack));
+    let completion_str = "^sleep";
+    let suggestions = completer.complete(completion_str, completion_str.len());
+    let expected: Vec<String> = vec!["sleep".into()];
+    match_suggestions(&expected, &suggestions);
+
+    let completion_str = "sleep";
+    let suggestions = completer.complete(completion_str, completion_str.len());
+    let expected: Vec<String> = vec!["sleep".into(), "^sleep".into()];
+    match_suggestions(&expected, &suggestions);
+}
```

Comment:
@ysthakur any idea how to make a similar test for windows?

Reply:
one thing that may help are commands that exist on all operating systems like `ping`. Some may be built-in and not work. I haven't tried but these also come to mind `echo`, `clear`, `mkdir`.

Reply:
What's the default path of those commands on windows?

Reply:
- ping - `c:\windows\system32\ping.exe`
- mkdir = cmd.exe built in so it's `"cmd /c 'mkdir /?'"`
- clear doesn't exist on windwos, my mistake. the closest thing is `"cmd /c 'cls /?'"`

Reply:
In case these executables aren't always in the same folders every time, another option would be to take one of the existing files in [the completions fixtures](https://github.com/nushell/nushell/tree/main/tests/fixtures/completions), make it executable, and add its folder to the PATH. We'd need to document clearly that the file is supposed to be marked executable, though

Reply:
Thanks for the advice from both of you. I went for the fixtures way, as it results in shorter test code.